pymc3/step_methods/smc_ABC.py

```
@@ -0,0 +1,418 @@
+        independent Markov Chains. Defaults to 5000.
+    step : :class:`SMC`
+        SMC initialization object
+    progressbar : bool
+        Flag for displaying a progress bar
+    model : pymc3 Model
+        optional if in `with` context
+    random_seed : int
+        random seed
+    """
+    model = modelcontext(model)
+
+    if random_seed != -1:
+        np.random.seed(random_seed)
+
+    stage = 0
+    variables = model.vars
+    discrete = np.concatenate(
+        [[v.dtype in pm.discrete_types] * (v.dsize or 1) for v in variables])
+    any_discrete = discrete.any()
+    all_discrete = discrete.all()
+    prior_logp = theano.function(variables, model.varlogpt)
+    simulator = model.observed_RVs[0]
+    function = simulator.distribution.function
+    observed_sum_stat = get_sum_stats(
+        simulator.observations,
+        sum_stat=step.sum_stat)
+    epsilon = np.inf
+    distance_list = []
+    pm._log.info('Using {} as distance metric'.format(step.distance_metric))
+    pm._log.info('Using {} as summary statistic'.format(step.sum_stat))
+
+    if step.epsilons is None:
+        epsilon_list = []
+
+    else:
+        step.epsilons.append(step.min_epsilon)
+        epsilon_list = step.epsilons
+
+    while epsilon > step.min_epsilon:
+        if stage == 0:
+            pm._log.info('Sample initial stage: ...')
+            posterior = _initial_population(draws, model, variables)
+            weights = np.ones(draws) / draws
+        else:
+            weights = calc_weights(un_weights)
+
+        # resample based on plausibility weights (selection)
+        resampling_indexes = np.random.choice(
+            np.arange(draws), size=draws, p=weights)
+        posterior = posterior[resampling_indexes]
+        # compute proposal distribution based on weights
+        try:
+            covariance = _calc_covariance(posterior, weights)
+        except ValueError:
+            pm._log.info('Could not compute covariance matrix')
+            break
```

raise an informative error instead of failing silently